activation/post.go

```diff
@@ -223,55 +246,49 @@ func (mgr *PostSetupManager) StartSession(ctx context.Context, opts PostSetupOpt
 	mgr.logger.With().Info("post setup completed",
 		log.String("node_id", mgr.id.String()),
 		log.String("commitment_atx", mgr.commitmentAtxId.String()),
-		log.String("data_dir", opts.DataDir),
-		log.String("num_units", fmt.Sprintf("%d", opts.NumUnits)),
+		log.String("data_dir", cfg.opts.DataDir),
+		log.String("num_units", fmt.Sprintf("%d", cfg.opts.NumUnits)),
 		log.String("labels_per_unit", fmt.Sprintf("%d", mgr.cfg.LabelsPerUnit)),
-		log.String("provider", fmt.Sprintf("%d", opts.ComputeProviderID)),
+		log.String("provider", fmt.Sprintf("%d", cfg.opts.ComputeProviderID)),
 	)
 	mgr.state = PostSetupStateComplete
 	return nil
 }
 
-func (mgr *PostSetupManager) prepareInitializer(ctx context.Context, opts PostSetupOpts) error {
+// InitializerConfig builds the config to be used by initialization in calls to
+// StartSession. Having this function be separate from StartSession provides a
+// means to understand if the post configuration is valid before kicking off a
+// very long running task (StartSession can take hours to complete).
```

Comment:
I don't understand this reasoning. If before the post configuration was invalid `prepareInitializer` would return an error and `StartSession` would stop immediately. 

Can you explain the reasoning as to why this has now to be called explicitly before calling `StartSession`?

Reply:
Previously with `prepareInitializer` being called in `StartSession` and since`StartSession` can take days to complete we would call it in a go-routine, so even if it stopped immediately any errors would not be available to the calling code that initiated the routine. The calling code could wait for the result from the go-routine but we wouldn't want to make the API call take day's to return in the case that initialization was successful.

So then the only option I see is to wait for "some" time and if we don't get a failure from the routine in that time, assume that initialization has started successfully. But the problem with that is:
* It's going to be slow, because we'll need to choose a wait time long enough that all machines have completed executing `prepareInitializer` before the time expires.
* It's not reliable since we can't guarantee that in all cases machines will complete running `prepareInitializer` before the time expires.

Alternatively, and what I think is the best option is that we change the api to return a stream that callers listen to, but that I guess is a much bigger change, since we'd need to update clients.

Reply:
> so even if it stopped immediately any errors would not be available to the calling code that initiated the routine.

I can follow that reasoning, but I think a synchronous "ValidatePostOps" method that returns a non-nil error if the provided options are incorrect would solve that issue. In my opinion this method not return a new instance of initializer and `commitmentAtx` can be part of the `PostSetupOpts`. It would make perfect sense to do that now, with the changes to how it is selected and now that any existing `Atx` is valid to be chosen for the ATX I don't see the need to treat it differently from all the other options.

> Alternatively, and what I think is the best option is that we change the api to return a stream that callers listen to, but that I guess is a much bigger change, since we'd need to update clients.

This actually pretty much how our API works at the moment 🙂 -> see `SmesherService::PostSetupStatusStream` in `go-spacemesh/api/grpcserver/smesher_service.go`

Reply:
I don't have a full understanding of what you are suggesting with regards to the `commitmentAtx` but that change seems to be out of scope with regards to not failing silently.

That aside the reason for returning an intializer is explained here - https://github.com/spacemeshos/go-spacemesh/pull/4319/files/b14c55ae32e6de102c014cb0f164ededec95828f#r1175539280

>>Alternatively, and what I think is the best option is that we change the api to return a stream that callers listen to, but that I guess is a much bigger change, since we'd need to update clients.

>This actually pretty much how our API works at the moment slightly_smiling_face -> see SmesherService::PostSetupStatusStream in go-spacemesh/api/grpcserver/smesher_service.go

Hehe, lol. I was not aware of the `PostSetupStatusStream` but i did initially mention the approach of passing all the feedback through a stream to the caller, however @dshulyak suggested going for the validate up front approach, so that is the approach I took.

Reply:
> however @dshulyak suggested going for the validate up front approach, so that is the approach I took.

I'm not opposed to this approach. I'm just nitpicking the implementation 🙂.

I would expect a method call to the validate the smeshing config in `cmd/node/node.go` that `log.Fatals` if the provided config is invalid (somewhere around here: https://github.com/spacemeshos/go-spacemesh/blob/develop/cmd/node/node.go#L816-L823).

Right now the node crashes (due to a panic) if `StartSmeshing` in fails. This could be due because of an invalid config during startup but also due to the node being instructed via GRPC, where imo when an error happens it should be communicated back to smapp and not cause a crash.

Reply:
The current solution seems to be functionally equivalent to what I proposed in my previous comment so you can ignore it 🙂